### PR TITLE
Add metrics collection to the framework

### PR DESCRIPTION
**Pull request by shreyb**
This PR adds lightly-wrapped prometheus metrics collection to the code, as well as adding a cherrypy webserver to serve those metrics (and other future functions).  The metrics are served by default on port 8000, at /metrics, but this is all configurable.  These metrics can then be collected in plaintext either by a prometheus server, or some other metrics collection service.

It is also possible to run the decisionengine without the webserver, by passing the `--no-webserver` flag, or by changing the environment of the systemd unit to pass that flag into the start command.

### CODE REVIEW

**Comment by retzkek** [comment]
@shreyb can you post an example of the currently published metrics? Just a raw scrape of /metrics.

**Comment by shreyb** [comment]
> @shreyb can you post an example of the currently published metrics? Just a raw scrape of /metrics.

Yep - here you go:

```
$ de-client --metrics
# HELP de_client_status_duration_seconds Multiprocess metric
# TYPE de_client_status_duration_seconds summary
de_client_status_duration_seconds_count 0.0
de_client_status_duration_seconds_sum 0.0
# HELP de_client_print_product_duration_seconds Multiprocess metric
# TYPE de_client_print_product_duration_seconds summary
de_client_print_product_duration_seconds_count 0.0
de_client_print_product_duration_seconds_sum 0.0
# HELP de_client_metrics_duration_seconds Multiprocess metric
# TYPE de_client_metrics_duration_seconds summary
de_client_metrics_duration_seconds_count 81.0
de_client_metrics_duration_seconds_sum 0.11079557938501239
# HELP de_client_start_channel_duration_seconds Multiprocess metric
# TYPE de_client_start_channel_duration_seconds summary
de_client_start_channel_duration_seconds_count{channel_name="Nersc"} 1.0
de_client_start_channel_duration_seconds_sum{channel_name="Nersc"} 0.7267125048674643
de_client_start_channel_duration_seconds_count{channel_name="test_channel"} 1.0
de_client_start_channel_duration_seconds_sum{channel_name="test_channel"} 1.3527377960272133
# HELP de_workers_total Multiprocess metric
# TYPE de_workers_total gauge
de_workers_total{pid="2102"} 1.0
de_workers_total{pid="2126"} 0.0
# HELP de_source_run_seconds Multiprocess metric
# TYPE de_source_run_seconds summary
de_source_run_seconds_count{channel_name="test_channel",source_name="SourceNOP"} 4666.0
de_source_run_seconds_sum{channel_name="test_channel",source_name="SourceNOP"} 4.778663814533502
# HELP de_transform_run_seconds Multiprocess metric
# TYPE de_transform_run_seconds summary
de_transform_run_seconds_count{channel_name="test_channel",transform_name="TransformNOP"} 4665.0
de_transform_run_seconds_sum{channel_name="test_channel",transform_name="TransformNOP"} 79.57181591587141
# HELP de_logicengine_run_seconds Multiprocess metric
# TYPE de_logicengine_run_seconds summary
de_logicengine_run_seconds_count{channel_name="test_channel",logicengine_name="LogicEngine"} 4665.0
de_logicengine_run_seconds_sum{channel_name="test_channel",logicengine_name="LogicEngine"} 90.91386365192011
# HELP de_publisher_run_seconds Multiprocess metric
# TYPE de_publisher_run_seconds summary
de_publisher_run_seconds_count{channel_name="test_channel",publisher_name="PublisherNOP"} 4665.0
de_publisher_run_seconds_sum{channel_name="test_channel",publisher_name="PublisherNOP"} 9.050738890189677
# HELP de_source_last_acquire_timestamp_seconds Multiprocess metric
# TYPE de_source_last_acquire_timestamp_seconds gauge
de_source_last_acquire_timestamp_seconds{channel_name="test_channel",pid="2126",source_name="SourceNOP"} 1.6324228903925512e+09
# HELP de_transform_last_run_timestamp_seconds Multiprocess metric
# TYPE de_transform_last_run_timestamp_seconds gauge
de_transform_last_run_timestamp_seconds{channel_name="test_channel",pid="2126",transform_name="TransformNOP"} 1.6324228904440765e+09
# HELP de_logicengine_last_run_timestamp_seconds Multiprocess metric
# TYPE de_logicengine_last_run_timestamp_seconds gauge
de_logicengine_last_run_timestamp_seconds{channel_name="test_channel",logicengine_name="LogicEngine",pid="2126"} 1.6324228904877343e+09
# HELP de_publisher_last_run_timestamp_seconds Multiprocess metric
# TYPE de_publisher_last_run_timestamp_seconds gauge
de_publisher_last_run_timestamp_seconds{channel_name="test_channel",pid="2126",publisher_name="PublisherNOP"} 1.6324228905370376e+09
# HELP de_channel_state Multiprocess metric
# TYPE de_channel_state gauge
de_channel_state{channel_name="test_channel",pid="2126"} 3.0
```

File: .editorconfig [hunk 1]
```diff
@@ -11,7 +11,7 @@ charset = utf-8
 [*.py]
 indent_style = space
 indent_size = 4
-max_line_length = 120
+max_line_length = 80
```
Comment (jcpunk) [inline .editorconfig:14]: I believe this is in error.

Comment (shreyb) [reply .editorconfig:14]: It is - I'll change it back right away

File: src/decisionengine/framework/taskmanager/TaskManager.py [hunk 2]
```diff
@@ -18,11 +18,93 @@
 from decisionengine.framework.modules.Transform import Transform
 from decisionengine.framework.taskmanager.module_graph import ensure_no_circularities
 from decisionengine.framework.taskmanager.ProcessingState import ProcessingState, State
+from decisionengine.framework.util.metrics import *
```
Comment (jcpunk) [inline src/decisionengine/framework/taskmanager/TaskManager.py:21]: Generally `import *` makes me nervous....

Can you add a `__all__` to `decisionengine.framework.util.metrics` to help limit the risk here?

File: src/decisionengine/framework/engine/DecisionEngine.py [hunk 3]
```diff
@@ -570,13 +651,16 @@ def main(args=None):
     If args is a list, it will be used instead of sys.argv (for unit testing)
     """
     options = parse_program_options(args)
+    _check_metrics_env(options)
     global_config_dir = policies.global_config_dir()
     global_config, channel_config_loader = _get_de_conf_manager(
         global_config_dir, policies.channel_config_dir(), options
     )
     try:
+        # app.run()
```
Comment (jcpunk) [inline src/decisionengine/framework/engine/DecisionEngine.py:660]: should this be commented out?

File: src/decisionengine/framework/taskmanager/TaskManager.py [hunk 4]
```diff
@@ -319,23 +437,24 @@ def run_source(self, src):
...
+                SOURCE_ACQUIRE_GAUGE.labels(self.name, src.name).set_to_current_time()
                 if data:
                     t = time.time()
                     header = datablock.Header(self.data_block_t0.taskmanager_id, create_time=t, creator=src.module)
                     self.logger.info(f"Src {src.name} header done")
                     self.data_block_put(data, header, self.data_block_t0)
                     self.logger.info(f"Src {src.name} data block put done")
                 else:
-                    self.logger.warning(f"Src {src.name} acquire returned no data")
+                    self.logger.warning(f"Src {src.name} acquire retuned no data")
```
Comment (jcpunk) [inline src/decisionengine/framework/taskmanager/TaskManager.py:457]: I believe the spelling change is in error here....

File: src/decisionengine/framework/taskmanager/TaskManager.py (hunk 2 above)
Comment (shreyb) [reply src/decisionengine/framework/taskmanager/TaskManager.py:21]: Sure thing.

File: src/decisionengine/framework/engine/DecisionEngine.py (hunk 3 above)
Comment (shreyb) [reply src/decisionengine/framework/engine/DecisionEngine.py:660]: Actually, during the massive rebase I just did, I thought that was leftover from the main branch, but I'm mistaken on that.  I'll get rid of that.

File: src/decisionengine/framework/taskmanager/TaskManager.py (hunk 2 above)
Comment (shreyb) [reply src/decisionengine/framework/taskmanager/TaskManager.py:21]: Actually, would it be better to add an __all__ there, or just import exactly what I need (which right now _is_ *, but could well change in the future)?

Comment (jcpunk) [reply src/decisionengine/framework/taskmanager/TaskManager.py:21]: Lets go with both.  Add an `__all__` and import only the specifics that you need.  Generally I find an `__all__` lets me feel better about adding local functions I don't intend to export.

**Comment by jcpunk** [comment]
Can you squash this down to a smaller set of commits?

**Comment by shreyb** [comment]
> Can you squash this down to a smaller set of commits?

@jcpunk  Done.

**Comment by shreyb** [comment]
When I've resolved the rebase errors, I'll squash those down to one commit and re-push this branch.

**Comment by shreyb** [comment]
> When I've resolved the rebase errors, I'll squash those down to one commit and re-push this branch.

Done.

**Comment by shreyb** [comment]
Would someone be able to take a look at this by any chance? This PR has been open for a while, and I've had to rebase twice since it's been ready for review, which, on a larger PR like this, is risky.  Thanks!

**Comment by jcpunk** [comment]
I don't see any obvious show stoppers.

**Review by vitodb — APPROVED** [review]
By looking at the code an test results all seems to be good.
I' approving this PR also based on feedback provided by Pat that had a close look at this PR since the beginning.

**Pull request merged**